src/hotspot/share/code/compressedStream.hpp

```diff
@@ -111,4 +112,122 @@ class CompressedWriteStream : public CompressedStream {
   }
 };
 
+class CompressedSparseData : public ResourceObj {
+protected:
+  int     _position; // current byte offset
+  size_t  _bit_position {0}; // current bit offset
+
+public:
+  CompressedSparseData(int position = 0) {
+    _position = position;
+  }
+};
+
+// Modified compression algorithm for a data set in which a significant part of the data is null
+class CompressedSparseDataReadStream : public CompressedSparseData {
+protected:
+  const u_char* _buffer;
+public:
+  CompressedSparseDataReadStream(const u_char* buffer, int position) : CompressedSparseData(position) {
+    _buffer = buffer;
+  }
+
+  const u_char* buffer() const { return _buffer; }
+
+  void set_position(int pos) {
+    _bit_position = 0;
+    _position = pos;
+  }
+
+  jboolean read_bool()       { return read_int(); }
+  jbyte    read_byte()       { return read_int(); }
+  jint     read_signed_int() { return UNSIGNED5::decode_sign(read_int()); }
+  jint     read_int();
+  jdouble  read_double() {
+    jint h = reverse_bits(read_int());
+    jint l = reverse_bits(read_int());
+    return jdouble_cast(jlong_from(h, l));
+  }
+  jlong    read_long() {
+    jint low  = read_signed_int();
+    jint high = read_signed_int();
+    return jlong_from(high, low);
+  }
+
+protected:
+  bool read_zero();
+  uint8_t read_byte_impl();
+  inline u_char read()       { return _buffer[_position++]; }
+};
+
+class CompressedSparseDataWriteStream : public CompressedSparseData {
+protected:
+  u_char* _buffer;
+public:
+  CompressedSparseDataWriteStream(int initial_size) : CompressedSparseData() {
+    _buffer   = NEW_RESOURCE_ARRAY(u_char, initial_size);
+    _size     = initial_size;
+    _buffer[0] = 0;
+  }
+
+  const u_char* buffer() const { return _buffer; }
+
+  void write_bool(jboolean value)   { write_int(value ? 1 : 0); }
+  void write_byte(jbyte value)      { write_int(value); }
+  void write_signed_int(jint value) { write_int(UNSIGNED5::encode_sign(value)); }
+  void write_int(juint value);
+  void write_double(jdouble value)  {
+    juint rh = reverse_bits(high(jlong_cast(value)));
+    juint rl = reverse_bits(low( jlong_cast(value)));
+    write_int(rh);
+    write_int(rl);
+  }
+  void write_long(jlong value)      {
+    write_signed_int(low(value));
+    write_signed_int(high(value));
+  }
+
+  void align() {
+    if (_bit_position > 0) {
+      // start a new byte
+      _position++;
+      _bit_position = 0;
+    }
+  }
+
+public:
+  // Start grouped data. Return a byte offset position in the stream where grouped data begins
+  int start_scope() {
+    align(); // a side effect!
+    return _position;
+  }
+
+  // Roll the stream state back to the given position
+  void roll_back(int pos) {
+    _position = pos;
+    _bit_position = 0;
+    assert(_position < _size, "set_position is only used for rollback");
```

Comment:
Should we change the assert and move it to the beginning of the function?
```
assert(pos <= _position, "new position must be rollback the current position"
```

Reply:
right. thank you